### ISSUE TITLE
DenseNet on ImageNet 

### ISSUE THREAD

**Issue by AlexandreBriot** [liuzhuang13/DenseNet#7]
I've just read your paper which is really interesting.
I was wondering whether you tried learning a DenseNet version on ImageNet ?
Thank you

**Comment by liuzhuang13** [comment]
Thanks for your interest in DenseNet.

We are experimenting on ImageNet with different model sizes. Right now we have some preliminary results (relatively small models), which is shown in the figures below.

![11 pic_hd](https://cloud.githubusercontent.com/assets/8370623/19392779/ccff53ce-9264-11e6-9ce4-4ebc77cd716a.jpg)
![12 pic_hd](https://cloud.githubusercontent.com/assets/8370623/19392788/d6f95de8-9264-11e6-8041-2709fec1a837.jpg)

As shown in the figure, DenseNet with the same amount of parameters of computation (measured in #flops) as ResNet performs better. When all the models are finished we'll update the paper and Readme with ImageNet results.

The DenseNet architecture used in ImageNet is different from what we used in CIFAR and SVHN dataset. The differences are listed below:

1. The major difference is we used "bottleneck structure", which is inspired by the ResNet paper. In each layer, before producing new feature maps through 3*3 convolution on previous layers' feature maps, a 1*1 convolution with output size 4*growthRate is performed.

2. In transition layers we halved the number of feature maps.

3. We use 4 dense blocks, and the depth in each dense block is different. This also follows the design of ResNet on ImageNet.

**Comment by AlexandreBriot** [comment]
Thanks for your answer.
That sounds very promising !

**Comment by chengchaoshen** [comment]
Why is there so large gap on parameters between ResNet-101 and ResNet-110 in the following table ?
![image](https://cloud.githubusercontent.com/assets/22129750/19506568/2b387734-95ff-11e6-90c3-51f234e86f7d.png)
Thank you.

**Comment by baiyancheng20** [comment]
Great results. 
When will you release the prototxt file for imagenet?

**Comment by liuzhuang13** [comment]
@baiyancheng20 sorry this is trained using torch. If you want to use them we can give you torch model definitions (or pre-trained models later) first.

Model definition here:
[densenet-imagenet.txt](https://github.com/liuzhuang13/DenseNet/files/546310/densenet-imagenet.txt)


After we get the full results we'll include imagenet models in both Torch and caffe repos.

**Comment by baiyancheng20** [comment]
@liuzhuang13 Thank you for sharing the code. Densenet is a very interesting work. I will try to use the code for Cifar to train on Imagenet Dataset. Thanks a lot.

**Comment by liuzhuang13** [comment]
@baiyancheng20 Thanks for your interest. In order to get better performance, you may want to adapt the caffe code for CIFAR a little bit according to the differences I listed above. For more detail you can refer to the torch code.

**Comment by livenletdie** [comment]
> Model definition here:
> densenet-imagenet.txt

The network model in the 'densenet-imagenet.txt' seems to be different from the paper. In the paper, DenseNet 169 has four dense blocks of size {6, 12, 32, 32} but the file has {6, 12, 48, 16}. Does that make a big difference? I am trying to train the network for Imagenet but convergence curve after the first 32 epochs does not look great (I am using fb.resnet.torch setup and just specified this network type via the -netType).  

Thanks
Ganesh

**Comment by liuzhuang13** [comment]
Sorry, the file was wrong, it was probably an older version. I'll correct it. Sorry but I couldn't remember whether this would make a big difference. 

Also, there was pretrained models available in the readme page, in case your purpose is just to use a pretrained model.

**Comment by livenletdie** [comment]
Thanks for the reply. No problem at all -- just wanted to confirm. 

Thank you for uploading the pre-trained models. They have been very helpful but I did want to train a model for a different study I was doing.

**Comment by karaszka** [comment]
Hi! I'm trying to train DenseNet-121-BC on the ImageNet (my own implementation) and I am just wondering weather the training curves I'm getting are any close to what it looked like for you. It would be great if you could share some of them for comparison or give me your opinion on mine results.

![image](https://cloud.githubusercontent.com/assets/11075451/22103972/7ae37f6e-de3d-11e6-92f1-d7126bebf5a7.png)
![selection_017](https://cloud.githubusercontent.com/assets/11075451/22103986/8681b35e-de3d-11e6-9987-ec1a93d0bb14.png)

In this setup one epoch lasts for roughly 25.6k iterations, so above you can see around 10 epochs (I'm using just one GPU for trining), those are the params that I'm using:

![selection_021](https://cloud.githubusercontent.com/assets/11075451/22104067/da9d40e8-de3d-11e6-937e-68fe38d898ae.png)
![selection_022](https://cloud.githubusercontent.com/assets/11075451/22104097/fdda9880-de3d-11e6-9844-509eb87f8d46.png)

Thanks!

**Comment by livenletdie** [comment]
Yes it would be great if the authors could post their convergence curves -- I tried to train with fb.resnet.torch repo where I just replace netType to DenseNet but my initial training curve looked weird. It would helpful if I had a curve to compare to so that I will know if it is expected or I am doing something wrong. Thanks in advance for the help!

**Comment by nihalgoalla** [comment]
Could you post the prototxt files used for training DenseNet's in caffe?

It would be great to check and make some changes to it to experiment.

**Comment by liuzhuang13** [comment]
Hi, @nihalgoalla please check https://github.com/liuzhuang13/DenseNetCaffe (for CIFAR, without BC structure) and https://github.com/shicai/DenseNet-Caffe (for ImageNet).

**Comment by nihalgoalla** [comment]
Hey,

Thanks for the information.

But what I was looking for was the last layers you added, the loss and
accuracy layers, including the solver prototxt's.

If it is possible, could you share those with me.

Thanks,
Nihal Goalla,
IIT Delhi.

**Comment by liuzhuang13** [comment]
@nihalgoalla  At https://github.com/liuzhuang13/DenseNetCaffe, we have a solver prototxt file (for CIFAR training) and an example prototxt file that contains the last layers. Thanks

**Comment by rishabhiitbhu** [comment]
@liuzhuang13 Did you scale the ImageNet images to `[0,1]` before feeding to DenseNet?

**Comment by zakirbd** [comment]
Hi,
I tried to extract image features using DenseNet-121 which is pre-trained (ImageNet). What would be the shape of the output features?

**Comment by JieMEI1994** [comment]
Hi @liuzhuang13, thanks for your great work in dense net.
Comparing to resnet, I wonder why you choose concatenate but not the add function in original resnet.
hope to hear you soon.

**Comment by liuzhuang13** [comment]
Hi @JieMEI1994, I think this is discussed in the section 5 of the paper.

**Comment by guotong1988** [comment]
Thank you.
